bech32: Initial Version

matheusd opened the pull request:
This is a port from the github.com/btcsuite/btcutil/bech32 package.

matheusd commented:
@dajohi If you feel this is something important, I can include those as well. Can you point me to which tests specifically you ran?

dajohi commented:
https://github.com/bitcoin/bitcoin/blob/master/src/test/bech32_tests.cpp#L48-L61

davecgh commented on bech32/bech32.go:
```diff
@@ -13,9 +13,14 @@ const charset = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
+// requests), NOT on-chain addresses.
+//
+// Note that the returned data is 5-bit (base32) encoded.
+func DecodeNoLimit(bech string) (string, []byte, error) {
```
The function is still failing on length > 90

davecgh commented on bech32/bech32.go:
```diff
@@ -0,0 +1,263 @@
+// Copyright (c) 2017 The btcsuite developers
+// Use of this source code is governed by an ISC
```
Need to add The Decred devs since it has the new function now.

davecgh commented on bech32/error.go:
```diff
@@ -0,0 +1,81 @@
+package bech32
```
Needs license header.

matheusd commented:
Fixed the copyright and documentation issues.

Also, I've added a benchmark for number of allocations of an encode/decode cycle and refactored the functions to reduce it to the minimum (2 - one for each of encode and decode results).

davecgh commented on bech32/bech32.go:
```diff
@@ -0,0 +1,339 @@
+//
+// Note that the returned data is 5-bit (base32) encoded.
+func DecodeNoLimit(bech string) (string, []byte, error) {
+	// The maximum allowed length for a bech32 string is 90. It must also
```
Since this is in `DecodeNoLimit` it doesn't have a maximum.

davecgh commented on bech32/bech32.go:
```diff
@@ -0,0 +1,339 @@
+
+	// Account for the separator (0) between high and low bits of the HRP.
+	b := chk >> 25
+	chk = (chk&0x1ffffff)<<5 ^ 0
```
Since `x^0 == x`, I would suggest leaving the xor off here.

davecgh commented on bech32/bech32.go:
```diff
@@ -0,0 +1,339 @@
+		// A nil checksum is used during encoding, so assume all bytes are zero.
+		for v := 0; v < 6; v++ {
+			b := chk >> 25
+			chk = (chk&0x1ffffff)<<5 ^ 0
```
Same here.
